### PR TITLE
chore: improve batching performance with batched queries

### PR DESCRIPTION
**Pull request by prkpndy**
## Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #NA

## What is the new behavior?

-
-
-

### CODE REVIEW

**Review by Mohiiit — COMMENTED** [review]
## Code Review Summary

Overall, this is a well-designed batching optimization that should provide significant performance improvements (~10-20x reduction in HTTP requests). The architecture is clean with good separation of concerns.

### Inline Comments

I've left a few inline comments on specific areas - mostly questions about intended behavior and a suggestion for improved testability.

### Test Coverage Gaps ⚠️

The current testing relies heavily on integration tests with real Sepolia data, which is great for the happy path. However, there are significant gaps in test coverage that would be valuable to address:

**Error Handling & Retry Logic (High Priority)**
- No tests for retry exhaustion scenarios
- No tests for partial batch failures (some requests succeed, some fail)
- No tests for HTTP timeouts and network errors
- No tests for invalid JSON responses

**Edge Cases (Medium Priority)**
- Empty input collections (`[]`, `{}`)
- Single-item batches
- Batch boundary tests (99, 100, 101 items to verify chunking)
- Very large batches (1000+ items)

**Response Parsing (Medium Priority)**
- Malformed Felt hex strings
- Missing response IDs
- Responses with neither result nor error

**Specific Functions Missing Direct Tests:**
| Function | Gap |
|----------|-----|
| `send_batch_with_retry()` | All retry scenarios |
| `send_batch()` | HTTP failure handling |
| `process_storage_diffs_batched()` | Empty inputs, all-zero values |
| `process_deployed_contracts_and_replaced_classes_batched()` | Empty inputs, unchanged class hash |

Adding these tests would provide confidence in error handling and make future refactoring safer.

**Review by Mohiiit — COMMENTED** [review]
Overall, this is a well-designed batching optimization with clean architecture. I have a few questions about intended behavior and a suggestion for improved testability.

**Test Coverage Gaps:**
The current tests focus on the happy path with real Sepolia data. Consider adding tests for:
- Error handling paths (retry exhaustion, partial batch failures, network errors)
- Edge cases (empty inputs, batch boundaries at 99/100/101 items)
- Response parsing (malformed responses, ID validation)

Key functions missing direct tests: `send_batch_with_retry()`, `send_batch()`

File: orchestrator/src/compression/batch_rpc.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        block_id: BlockId,
+    ) -> Result<HashMap<(Felt, Felt), Felt>, BatchRpcError> {
+        if queries.is_empty() {
+            return Ok(HashMap::new());
+        }
+
+        let block_param = Self::block_id_to_param(&block_id);
+
+        // Chunk queries into batches
+        let chunks: Vec<Vec<(Felt, Felt)>> = queries.chunks(self.config.batch_size).map(|c| c.to_vec()).collect();
```
Comment (Mohiiit) [inline orchestrator/src/compression/batch_rpc.rs:143]: **Nit:** The `.collect()` here loads all batch results into memory before processing. For very large state updates, this could increase memory pressure. Is this intentional?

File: orchestrator/src/compression/batch_rpc.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        let responses = self.send_batch_with_retry(&requests).await?;
+
+        // Parse responses
+        let mut results = HashMap::new();
+        for (idx, (contract_addr, key)) in queries.iter().enumerate() {
+            let response = responses.get(&(idx as u64)).ok_or(BatchRpcError::MissingResponse(idx as u64))?;
+
+            let value = match response {
+                Ok(val) => Self::parse_felt_result(val)?,
+                Err(err) => {
```
Comment (Mohiiit) [inline orchestrator/src/compression/batch_rpc.rs:228]: Is this behavior intentional? Currently all RPC errors (including network failures) become `Felt::ZERO`. Should we distinguish "storage not found" (expected) from network failures (unexpected)?

File: orchestrator/src/compression/batch_rpc.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            let response = responses.get(&(idx as u64)).ok_or(BatchRpcError::MissingResponse(idx as u64))?;
+
+            let value = match response {
+                Ok(val) => Some(Self::parse_felt_result(val)?),
+                Err(err) => {
+                    if err.code == CONTRACT_NOT_FOUND_CODE {
+                        // Contract doesn't exist, this is expected
+                        None
+                    } else {
+                        // Unexpected error, log but return None to not block processing
```
Comment (Mohiiit) [inline orchestrator/src/compression/batch_rpc.rs:279]: Same question - is it intentional that all errors return `None`? Network failures would be silently treated as "contract doesn't exist."

File: orchestrator/src/compression/squash.rs [hunk 4]
```diff
@@ -185,189 +171,285 @@ impl StateDiffMap {
...
+    let pre_range_values = if storage_queries.is_empty() {
+        HashMap::new()
+    } else {
+        batch_client
+            .batch_get_storage_at(storage_queries, block_id)
+            .await
+            .map_err(|e| JobError::ProviderError(format!("Failed to batch get storage values: {}", e)))?
+    };
+
+    // Step 4: Build storage diffs by comparing values
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs:253]: Consider extracting the comparison logic (pre-range vs new value) into pure functions like `build_entries_for_existing_contract()`. This would allow unit testing without mocking RPC calls - important given the limited test coverage for edge cases.

File: orchestrator/src/compression/batch_rpc.rs (hunk 2 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs:228]: fair point. actually, we should return error in all cases since we are already checking for contract existence. it never returns storage not found (for non-existing address, it returns 0, which works for us). thanks for pointing it out!

File: orchestrator/src/compression/batch_rpc.rs (hunk 3 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs:279]: nopes. updated it to return None only when contract is not found

File: orchestrator/src/compression/batch_rpc.rs (hunk 1 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs:143]: umm, this is only holding the chunks, so basically vector of tuple (contract, storage). it wont be very big. tried to remove that but can't do it because chunks() method returns borrowed slices which don't satisfy Send bounds for async. have also added a comment for this

File: orchestrator/src/compression/squash.rs (hunk 4 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs:253]: done

**Review by prkpndy — COMMENTED** [review]
resolved pr comments

**Review by byteZorvin — COMMENTED** [review]
Starting review of batching performance improvements.

File: orchestrator/src/compression/batch_rpc.rs [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    ) -> Result<HashMap<u64, Result<serde_json::Value, JsonRpcErrorData>>, BatchRpcError> {
+        let body = serde_json::to_string(requests)?;
+
+        let response = self
+            .client
+            .post(self.rpc_url.as_str())
+            .header("Content-Type", "application/json")
+            .body(body)
+            .send()
+            .await?
```
Comment (byteZorvin) [inline orchestrator/src/compression/batch_rpc.rs:338]: The HTTP client has no timeout configured. Consider adding a timeout to prevent long-running or hanging requests from blocking indefinitely:
```rust
.timeout(Duration::from_secs(30))
```

File: orchestrator/src/compression/squash.rs [hunk 6]
```diff
@@ -1,32 +1,25 @@
...
+///
+/// This function uses batch JSON-RPC calls to efficiently fetch pre-range block
+/// storage values and class hashes, reducing the number of HTTP requests by ~10x.
 pub async fn squash(
     state_updates: Vec<&StateUpdate>,
     pre_range_block: Option<u64>,
-    provider: &Arc<JsonRpcClient<HttpTransport>>,
+    batch_client: &BatchRpcClient,
 ) -> Result<StateUpdate, JobError> {
+    info!("Squashing state updates");
```
Comment (byteZorvin) [inline orchestrator/src/compression/squash.rs]: Nit: Consider using debug for start and end messages, could be too verbose for production loads.

File: orchestrator/src/compression/batch_rpc.rs [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+        match block_id {
+            BlockId::Number(n) => serde_json::json!({ "block_number": n }),
+            BlockId::Hash(h) => serde_json::json!({ "block_hash": format!("{:#x}", h) }),
+            BlockId::Tag(tag) => match tag {
+                starknet_core::types::BlockTag::Latest => serde_json::json!("latest"),
+                starknet_core::types::BlockTag::L1Accepted => serde_json::json!("l1_accepted"),
+                starknet_core::types::BlockTag::PreConfirmed => serde_json::json!("pre_confirmed"),
+            },
+        }
+    }
```
Comment (byteZorvin) [inline orchestrator/src/compression/batch_rpc.rs]: This manual conversion can be removed. starknet_core::types::BlockId already implements Serialize with the exact same JSON output format (see starknet-core/src/types/serde_impls.rs).

You can simplify by using serde_json::to_value(block_id)? directly instead of this function.

File: orchestrator/src/compression/batch_rpc.rs [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+    /// Parse a Felt from JSON-RPC response value
+    fn parse_felt_result(value: &serde_json::Value) -> Result<Felt, BatchRpcError> {
+        let hex_str =
+            value.as_str().ok_or_else(|| BatchRpcError::InvalidResponse("Expected string for Felt".to_string()))?;
+
+        Felt::from_hex(hex_str)
+            .map_err(|e| BatchRpcError::InvalidResponse(format!("Invalid Felt hex '{}': {}", hex_str, e)))
+    }
```
Comment (byteZorvin) [inline orchestrator/src/compression/batch_rpc.rs:365]: This can also be done with starknet 
```
serde_json::from_value::<Felt>(value.clone())
    .map_err(|e| BatchRpcError::InvalidResponse(e.to_string()))```

**Review by byteZorvin — CHANGES_REQUESTED** [review]

File: orchestrator/src/compression/squash.rs [hunk 9]
```diff
@@ -39,14 +32,17 @@ pub async fn squash(
...
+
+    let state_diff = state_diff_map.get_state_diff_batched(pre_range_block, batch_client).await?;
 
     // Create the merged StateUpdate
     let mut merged_update = StateUpdate { block_hash, new_root, old_root, state_diff };
 
     // Sort the merged StateUpdate
     sort_state_diff(&mut merged_update);
 
+    info!("Squashed state updates");
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs]: I think the logging in this function could be changed to debug

File: orchestrator/src/compression/squash.rs [hunk 10]
```diff
@@ -185,43 +171,195 @@ impl StateDiffMap {
...
-    provider: &Arc<JsonRpcClient<HttpTransport>>,
-    pre_range_block_option: Option<u64>,
+/// Process storage diffs using batched RPC calls.
+///
+/// For each contract with storage changes:
+/// 1. Check if the contract existed at pre_range_block
+/// 2. For existing contracts, fetch all pre-range storage values
+/// 3. Filter out entries where the value hasn't changed
+/// 4. For new contracts (didn't exist), filter out zero values
+async fn process_storage_diffs_batched(
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs:181]: I think this function can be divided in further function and it would provide better readability + ability to test things better

File: orchestrator/src/compression/squash.rs [hunk 11]
```diff
@@ -185,43 +171,195 @@ impl StateDiffMap {
...
+                None
+            }
+        })
+        .collect();
+
+    if entries.is_empty() {
+        None
+    } else {
+        Some(ContractStorageDiffItem { address: contract_addr, storage_entries: entries })
+    }
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs]:
```suggestion
    (!entries.is_empty()).then_some(ContractStorageDiffItem {
    address: contract_addr,
    storage_entries: entries,
})
```

File: orchestrator/src/compression/squash.rs [hunk 12]
```diff
@@ -234,140 +372,212 @@ async fn process_deployed_contracts_and_replaced_classes(
...
+                    }],
+                }],
+                deployed_contracts: vec![],
+                declared_classes: vec![],
+                deprecated_declared_classes: vec![],
+                nonces: vec![],
+                replaced_classes: vec![],
+                migrated_compiled_classes: None,
+            },
+        };
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs]: maybe a fixture can be sued here?

File: orchestrator/src/compression/squash.rs [hunk 13]
```diff
@@ -234,140 +372,212 @@ async fn process_deployed_contracts_and_replaced_classes(
...
+    // ========================================================================
+    // Tests for pure comparison functions
+    // ========================================================================
+
+    /// Comprehensive test for build_storage_diff_item covering all edge cases:
+    /// - Existing contract: unchanged values filtered, changed values included
+    /// - New contract (no pre-range): zero values filtered, non-zero included
+    /// - Returns None when all entries filtered out
+    #[test]
+    fn test_build_storage_diff_item() {
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs:447]: nice test case indeed!

File: orchestrator/src/compression/squash.rs [hunk 14]
```diff
@@ -234,140 +372,212 @@ async fn process_deployed_contracts_and_replaced_classes(
...
-        ))
-    })
+    /// Comprehensive test for filter_changed_replaced_classes covering all edge cases:
+    /// - Unchanged class hash (filtered)
+    /// - Changed class hash (included)
+    /// - New contract with None prev (included)
+    /// - Contract missing from prev map (included)
+    /// - Empty input (empty output)
+    #[test]
+    fn test_filter_changed_replaced_classes() {
```
Comment (Mohiiit) [inline orchestrator/src/compression/squash.rs:515]: I think this would be very clean with rstest, good test cases tho!

**Review by Mohiiit — COMMENTED** [review]
good tests, no logic change requested, more on code practices!

File: orchestrator/src/compression/squash.rs (hunk 10 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs:181]: hmm.. I am not sure if we'll gain anything by breaking it down further. This function has a simple purpose that given a contract address, its storage entries and a map of storage entries in pre-range, it'll give me the diff for the contract. Have written test for it as well. breaking it down further will complicate things and make it verbose as well. wdyt?

File: orchestrator/src/compression/batch_rpc.rs (hunk 5 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs:338]: done

File: orchestrator/src/compression/squash.rs (hunk 6 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs]: done

File: orchestrator/src/compression/batch_rpc.rs (hunk 7 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs]: done

File: orchestrator/src/compression/batch_rpc.rs (hunk 8 above)
Comment (prkpndy) [reply orchestrator/src/compression/batch_rpc.rs:365]: done

File: orchestrator/src/compression/squash.rs (hunk 9 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs]: done

File: orchestrator/src/compression/squash.rs (hunk 11 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs]: done

File: orchestrator/src/compression/squash.rs (hunk 12 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs]: hmm, can do this but still to make the function generic, most of the values would needed to be passed as arguments only - block hash, roots, storage diffs, etc. also, it's being used only in 2 tests. would be an overkill imo 😅

File: orchestrator/src/compression/squash.rs (hunk 14 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs:515]: can do if we need to use fixtures

**Review by byteZorvin — APPROVED** [review]
GG !!

File: orchestrator/src/compression/squash.rs (hunk 10 above)
Comment (Mohiiit) [reply orchestrator/src/compression/squash.rs:181]: gain is not on logic side at all! it's more about having more tests and simpler functions so that it's easier to read and test

File: orchestrator/src/compression/squash.rs (hunk 14 above)
Comment (Mohiiit) [reply orchestrator/src/compression/squash.rs:515]: yes we should use that

File: orchestrator/src/compression/squash.rs (hunk 12 above)
Comment (Mohiiit) [reply orchestrator/src/compression/squash.rs]: I think we should follow the best practices nonetheless?

**Review by Mohiiit — APPROVED** [review]

File: orchestrator/src/compression/squash.rs (hunk 12 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs]: added helper functions to create state update

File: orchestrator/src/compression/squash.rs (hunk 14 above)
Comment (prkpndy) [reply orchestrator/src/compression/squash.rs:515]: no need for rstest. fixtures were getting too complicated for this. added helper functions
